File: frontend/src/services/workspaces.ts
import { apiRequest } from './client'

export interface Workspace {
  id: string
  user_id: string
  name: string
  description: string | null
  created_at: string
  updated_at: string
}

export interface WorkspaceCreatePayload {
  name: string
  description?: string
}

export interface WorkspaceUpdatePayload {
  name?: string
  description?: string
}

export function listWorkspaces(token: string): Promise<Workspace[]> {
  return apiRequest('/workspaces', { method: 'GET' }, token)
}

export function getWorkspace(token: string, id: string): Promise<Workspace> {
  return apiRequest(`/workspaces/${id}`, { method: 'GET' }, token)
}

export function createWorkspace(token: string, payload: WorkspaceCreatePayload): Promise<Workspace> {
  return apiRequest('/workspaces', { method: 'POST', body: JSON.stringify(payload) }, token)
}

export function updateWorkspace(
  token: string,
  id: string,
  payload: WorkspaceUpdatePayload,
): Promise<Workspace> {
  return apiRequest(`/workspaces/${id}`, { method: 'PATCH', body: JSON.stringify(payload) }, token)
}

export function deleteWorkspace(token: string, id: string): Promise<void> {
  return apiRequest(`/workspaces/${id}`, { method: 'DELETE' }, token)
}

File: backend/app/api/workspaces.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_accessible_workspace, get_current_user, get_owned_workspace
from app.database import get_db
from app.models import User, Workspace
from app.schemas.workspace import WorkspaceCreate, WorkspaceResponse, WorkspaceUpdate
from app.services import workspace as workspace_service

router = APIRouter(prefix="/workspaces")


@router.get("", response_model=list[WorkspaceResponse])
async def list_workspaces(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return await workspace_service.list_workspaces(db, current_user)


@router.post("", response_model=WorkspaceResponse, status_code=status.HTTP_201_CREATED)
async def create_workspace(
    data: WorkspaceCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return await workspace_service.create_workspace(db, current_user, data)


@router.get("/{workspace_id}", response_model=WorkspaceResponse)
async def get_workspace(workspace: Workspace = Depends(get_accessible_workspace)):
    return workspace


@router.patch("/{workspace_id}", response_model=WorkspaceResponse)
async def update_workspace(
    data: WorkspaceUpdate,
    workspace: Workspace = Depends(get_owned_workspace),
    db: AsyncSession = Depends(get_db),
):
    if data.name is None and data.description is None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="No fields to update")
    return await workspace_service.update_workspace(db, workspace, data)


@router.delete("/{workspace_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_workspace(
    workspace: Workspace = Depends(get_owned_workspace),
    db: AsyncSession = Depends(get_db),
):
    await workspace_service.delete_workspace(db, workspace)
